src/core/vectortile/qgsvectortilemvtencoder.cpp

```diff
@@ -0,0 +1,357 @@
+/***************************************************************************
+  qgsvectortilemvtencoder.cpp
+  --------------------------------------
+  Date                 : April 2020
+  Copyright            : (C) 2020 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsvectortilemvtencoder.h"
+
+#include "qgslinestring.h"
+#include "qgslogger.h"
+#include "qgsmultilinestring.h"
+#include "qgsmultipoint.h"
+#include "qgsmultipolygon.h"
+#include "qgspolygon.h"
+#include "qgsvectorlayer.h"
+#include "qgsvectortilemvtutils.h"
+
+
+//! Helper class for writing of geometry commands
+struct MVTGeometryWriter
+{
+  vector_tile::Tile_Feature *feature = nullptr;
+  int resolution;
+  double tileXMin, tileYMax, tileDX, tileDY;
+  QPoint cursor;
+
+  MVTGeometryWriter( vector_tile::Tile_Feature *f, int res, const QgsRectangle &tileExtent )
+    : feature( f )
+    , resolution( res )
+    , tileXMin( tileExtent.xMinimum() )
+    , tileYMax( tileExtent.yMaximum() )
+    , tileDX( tileExtent.width() )
+    , tileDY( tileExtent.height() )
+  {
+  }
+
+  void addMoveTo( int count )
+  {
+    feature->add_geometry( 1 | ( count << 3 ) );
+  }
+  void addLineTo( int count )
+  {
+    feature->add_geometry( 2 | ( count << 3 ) );
+  }
+  void addClosePath()
+  {
+    feature->add_geometry( 7 | ( 1 << 3 ) );
+  }
+
+  void addPoint( const QgsPoint &pt )
+  {
+    addPoint( mapToTileCoordinates( pt ) );
+  }
+
+  void addPoint( const QPoint &pt )
+  {
+    qint32 vx = pt.x() - cursor.x();
+    qint32 vy = pt.y() - cursor.y();
+
+    feature->add_geometry( ( vx << 1 ) ^ ( vx >> 31 ) );
+    feature->add_geometry( ( vy << 1 ) ^ ( vy >> 31 ) );
+
+    cursor = pt;
+  }
+
+  QPoint mapToTileCoordinates( const QgsPoint &p )
+  {
+    return QPoint( static_cast<int>( round( ( p.x() - tileXMin ) * resolution / tileDX ) ),
+                   static_cast<int>( round( ( tileYMax - p.y() ) * resolution / tileDY ) ) );
+  }
+};
+
+
+static QVector<QgsPoint> pointsFromLineString( const QgsLineString *ls )
```

Comment:
This is rather inefficient -- you're better accessing QgsLineString::xData and yData direct and iterating over that, building the encoded line string directly during iteration

Reply:
I'd go one step further here, and not return QVector<QgsPoint> here but instead directly build the encoded line string. otherwise you're constructing a QVector of QgsPoints which aren't really needed for every linestring, which will definitely impact performance.

Reply:
Okay, that was just lazy myself - fixed in d025079 now :-)

There is still one intermediate vector of points created (instead of two) while encoding - the reason is that we need to do de-duplication of points in the transformed integer coordinates (spec does not allow duplicate points) and the point count needs to be written before encoding points. But I am not really worried about performance of the encoder (at least not at this stage :-) )